prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -5467,7 +5496,41 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
                 (expr.field(field), stmts)
             }
             PlaceEncoding::ArrayAccess { base, index, array_elem_ty, array_len } => {
-                panic!("ArrayAccess {{ {:?}, {:?}, {:?}, {:?} }}", base, index, array_elem_ty, array_len);
+                let elem_ty_name = if let vir::Type::TypedRef(ref name) = array_elem_ty { name } else { unreachable!() };
+                let lookup_pure_ret = rust_type_name_to_vir_type(elem_ty_name)?;
+                let lookup_pure = self.encoder.encode_builtin_function_use(
+                    BuiltinFunctionKind::ArrayLookupPure {
+                        array_elem_ty: array_elem_ty.clone(),
+                        array_len,
+                        return_ty: lookup_pure_ret.clone(),
+                    }
+                );
+
+                let array_type = base.get_type().clone();
+                let lookup_res: vir::Expr = self.get_auxiliary_local_var("lookup_ret", array_elem_ty).into();
+                let lookup_res_val_int = lookup_res.clone().field(vir::Field::new("val_int", vir::Type::Int));
```

Comment:
this seems more manual than it needs to be, but i currently don't see how to do this better. it boils down to

 - the temporary variable seems to have to be type `Ref`, because the assignment to the final variable generated wants to copy from `.val_int`
 - that seems to mean that we need to pass `.val_int` to `lookup_pure`

Reply:
Yes, as long as pure functions don't use snapshots for the parameters we have to add the `.val_int`. The field creation with the hardcoded string can be replaced with a call to `Encoder::encode_value_field`.

Reply:
The temporary variable is of Viper type `Ref` because of `array_elem_ty`. By default we encode primitive types as heap-allocated values to simplify the encoding of creation of references. One of the future optimizations  (purification) should optimize cases like this.